### PR TITLE
Introduce a <model> loading process - Part 1

### PR DESCRIPTION
**Pull request by eddydas**
#### d4bf0dab90e9db4ab1c487c933d93321afc8d3fb
<pre>
Introduce a &lt;model&gt; loading process - Part 1
<a href="https://bugs.webkit.org/show_bug.cgi?id=243874">https://bugs.webkit.org/show_bug.cgi?id=243874</a>

Reviewed by Alex Christensen.

In order to allow parsing of complex 3D model formats with a custom (tight) sandbox,
introduce a separate process for loading/parsing &lt;model&gt;s.

This is part 1 of the effort, where we have added the new classes and messages.
Most of them are gated behind ENABLE(MODEL_PROCESS) that remains false on all platforms
for the time being.

In an upcoming PR, we will introduce new code in existing classes that will call into
these new classes, as well as flipping on the compilation flag on applicable platforms.
The actual ModelProcess XPC service will come in an upcoming PR as well.

The new classes and messages in this patch is to support launching of the new process.
Most of them are boilerplate code adapted from GPUProcess. There is a good amount of
duplicated code that warrants a refactoring down the road, but that is out of scope
for this current &lt;model&gt; tag effort.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp: Added.
(WebKit::ModelConnectionToWebProcess::create):
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
(WebKit::m_presentingApplicationAuditToken):
(WebKit::ModelConnectionToWebProcess::~ModelConnectionToWebProcess):
(WebKit::ModelConnectionToWebProcess::didClose):
(WebKit::ModelConnectionToWebProcess::createVisibilityPropagationContextForPage):
(WebKit::ModelConnectionToWebProcess::destroyVisibilityPropagationContextForPage):
(WebKit::ModelConnectionToWebProcess::configureLoggingChannel):
(WebKit::ModelConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::ModelConnectionToWebProcess::logger):
(WebKit::ModelConnectionToWebProcess::didReceiveInvalidMessage):
(WebKit::ModelConnectionToWebProcess::lowMemoryHandler):
(WebKit::ModelConnectionToWebProcess::dispatchMessage):
(WebKit::ModelConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h: Added.
(WebKit::ModelConnectionToWebProcess::connection):
(WebKit::ModelConnectionToWebProcess::messageReceiverMap):
(WebKit::ModelConnectionToWebProcess::modelProcess):
(WebKit::ModelConnectionToWebProcess::webProcessIdentifier const):
(WebKit::ModelConnectionToWebProcess::sessionID const):
(WebKit::ModelConnectionToWebProcess::webProcessIdentity const):
(WebKit::ModelConnectionToWebProcess::objectCountForTesting):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcess.cpp: Added.
(WebKit::ModelProcess::ModelProcess):
(WebKit::ModelProcess::didReceiveMessage):
(WebKit::ModelProcess::createModelConnectionToWebProcess):
(WebKit::ModelProcess::removeModelConnectionToWebProcess):
(WebKit::ModelProcess::connectionToWebProcessClosed):
(WebKit::ModelProcess::shouldTerminate):
(WebKit::ModelProcess::canExitUnderMemoryPressure const):
(WebKit::ModelProcess::tryExitIfUnusedAndUnderMemoryPressure):
(WebKit::ModelProcess::tryExitIfUnused):
(WebKit::ModelProcess::lowMemoryHandler):
(WebKit::ModelProcess::initializeModelProcess):
(WebKit::ModelProcess::prepareToSuspend):
(WebKit::ModelProcess::processDidResume):
(WebKit::ModelProcess::resume):
(WebKit::ModelProcess::webProcessConnection const):
(WebKit::ModelProcess::webProcessConnectionCountForTesting):
(WebKit::ModelProcess::addSession):
(WebKit::ModelProcess::removeSession):
* Source/WebKit/ModelProcess/ModelProcess.h: Added.
(WebKit::ModelProcess::applicationVisibleName const):
* Source/WebKit/ModelProcess/ModelProcess.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.h: Added.
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.serialization.in: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/Shared/ModelProcessConnectionParameters.h: Added.
* Source/WebKit/Shared/ModelProcessConnectionParameters.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp: Added.
(WebKit::ModelProcessProxy::create):
(WebKit::ModelProcessProxy::ModelProcessProxy):
(WebKit::ModelProcessProxy::getLaunchOptions):
(WebKit::ModelProcessProxy::connectionWillOpen):
(WebKit::ModelProcessProxy::processWillShutDown):
(WebKit::ModelProcessProxy::createModelProcessConnection):
(WebKit::ModelProcessProxy::modelProcessExited):
(WebKit::ModelProcessProxy::processIsReadyToExit):
(WebKit::ModelProcessProxy::addSession):
(WebKit::ModelProcessProxy::removeSession):
(WebKit::ModelProcessProxy::terminateForTesting):
(WebKit::ModelProcessProxy::webProcessConnectionCountForTesting):
(WebKit::ModelProcessProxy::didClose):
(WebKit::ModelProcessProxy::didReceiveInvalidMessage):
(WebKit::ModelProcessProxy::didFinishLaunching):
(WebKit::ModelProcessProxy::updateProcessAssertion):
(WebKit::ModelProcessProxy::sendPrepareToSuspend):
(WebKit::ModelProcessProxy::sendProcessDidResume):
(WebKit::ModelProcessProxy::didCreateContextForVisibilityPropagation):
(WebKit::ModelProcessProxy::didBecomeUnresponsive):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h: Added.
* Source/WebKit/UIProcess/Model/ModelProcessProxy.messages.in: Added.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createModelProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
In addition to the newly added classes, also added the existing
GPUProcessConnectionInfo.serialization.in file so that it can be directly
edited from Xcode.

* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp: Added.
(WebKit::getModelProcessConnectionParameters):
(WebKit::ModelProcessConnection::create):
(WebKit::ModelProcessConnection::ModelProcessConnection):
(WebKit::ModelProcessConnection::~ModelProcessConnection):
(WebKit::ModelProcessConnection::auditToken):
(WebKit::ModelProcessConnection::invalidate):
(WebKit::ModelProcessConnection::didClose):
(WebKit::ModelProcessConnection::didReceiveInvalidMessage):
(WebKit::ModelProcessConnection::dispatchMessage):
(WebKit::ModelProcessConnection::dispatchSyncMessage):
(WebKit::ModelProcessConnection::didInitialize):
(WebKit::ModelProcessConnection::waitForDidInitialize):
(WebKit::ModelProcessConnection::createVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::destroyVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::configureLoggingChannel):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h: Added.
(WebKit::ModelProcessConnection::connection):
(WebKit::ModelProcessConnection::messageReceiverMap):
(WebKit::ModelProcessConnection::Client::modelProcessConnectionDidClose):
(WebKit::ModelProcessConnection::addClient):
(WebKit::ModelProcessConnection::removeClient):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.messages.in: Added.
* Source/WebKit/WebProcess/Model/ModelProcessConnectionInfo.h: Added.
* Source/WebKit/WebProcess/Model/ModelProcessConnectionInfo.serialization.in: Added.

The following files are modified trivially to accomodate for source unification reshuffling:
* Source/WebCore/Modules/push-api/PushStrategy.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:

Canonical link: <a href="https://commits.webkit.org/273840@main">https://commits.webkit.org/273840@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0deb8b649fc30633b6de6deb45f2c6a70e22dcb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39398 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32928 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18176 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31497 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13222 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32480 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11566 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/36863 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11576 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40646 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33101 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37495 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11855 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9670 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35627 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12254 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4770 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12751 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/DerivedSources-input.xcfilelist [hunk 1]
```diff
@@ -114,6 +114,10 @@ $(PROJECT_DIR)/GPUProcess/webrtc/RemoteMediaRecorder.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteMediaRecorderManager.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in
 $(PROJECT_DIR)/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.messages.in
+$(PROJECT_DIR)/ModelProcess/ModelConnectionToWebProcess.messages.in
+$(PROJECT_DIR)/ModelProcess/ModelProcess.messages.in
+$(PROJECT_DIR)/ModelProcess/ModelProcessCreationParameters.serialization.in
+$(PROJECT_DIR)/ModelProcess/mac/com.apple.WebKit.ModelProcess.sb.in
```
Comment (achristensen07) [inline Source/WebKit/DerivedSources-input.xcfilelist]: This file does not exist.

File: Source/WebKit/DerivedSources.make [hunk 2]
```diff
@@ -28,6 +28,8 @@ VPATH = \
     $(WebKit2)/GPUProcess/media \
     $(WebKit2)/GPUProcess/media/ios \
     $(WebKit2)/GPUProcess/webrtc \
+    $(WebKit2)/ModelProcess \
+    $(WebKit2)/ModelProcess/mac \
```
Comment (achristensen07) [inline Source/WebKit/DerivedSources.make]: There's nothing in this directory.  Could you remove this?  Also, let's put stuff in a directory named "cocoa" instead.

File: Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+#include <wtf/CompletionHandler.h>
+#include <wtf/LogInitialization.h>
+#include <wtf/MachSendRight.h>
+
+#if PLATFORM(IOS_FAMILY)
+#include <wtf/spi/darwin/XPCSPI.h>
+#endif
+
+#if PLATFORM(COCOA)
+#include <wtf/BlockPtr.h>
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp]: This isn't used.  Same with wtf/spi/darwin/XPCSPI.h

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 55;
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: This should be reverted.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -17972,10 +18116,11 @@
...
...PLATFORM_NAME}\" == iosmac ]]; then\n    ln -sfhv \"Versions/Current/XPCServices\" \"${BUILT_PRODUCTS_DIR}/WebKit.framework/XPCServices\";\nfi\n\nif [[ \"${DEPLOYMENT_LOCATION}\" == \"YES\" ]]; then\n    exit\nfi\n\nif [[ ${WK_PLATFORM_NAME} != \"macosx\" ]]; then\n    BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES=\"../..\"\nelse\n    BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES=\"../../../..\"\nfi\n\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.WebContent.xpc\" \"${SCRIPT_OUTPUT_FILE_0}\"\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.WebContent.CaptivePortal.xpc\" \"${SCRIPT_OUTPUT_FILE_1}\"\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.WebContent.Development.xpc\" \"${SCRIPT_OUTPUT_FILE_2}\"\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.Networking.xpc\" \"${SCRIPT_OUTPUT_FILE_3}\"\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.GPU.xpc\" \"${SCRIPT_OUTPUT_FILE_4}\"\nln -sFhv \"${BUILT_PRODUCTS_DIR_RELATIVE_PATH_FROM_XPC_SERVICES}/com.apple.WebKit.Model.xpc\" \"${SCRIPT_OUTPUT_FILE_5}\"\n";
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: This looks like it should be in the change that actually introduces the xpc service.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -19039,7 +19183,6 @@
 			isa = PBXTargetDependency;
 			platformFilters = (
 				ios,
-				xros,
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Revert these changes.

File: Source/WebKit/WebProcess/Model/ModelProcessConnection.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+#include "Connection.h"
+#include "MessageReceiverMap.h"
+#include "SharedMemory.h"
+#include <wtf/RefCounted.h>
+#include <wtf/WeakHashSet.h>
+#include <wtf/WeakPtr.h>
+#include <wtf/text/WTFString.h>
+
+namespace IPC {
+class Semaphore;
```
Comment (achristensen07) [inline Source/WebKit/WebProcess/Model/ModelProcessConnection.h]: This isn't used.

File: Source/WebKit/Shared/ModelProcessConnectionParameters.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(MODEL_PROCESS)
+
+#include "WebCoreArgumentCoders.h"
```
Comment (achristensen07) [inline Source/WebKit/Shared/ModelProcessConnectionParameters.h]: This is almost certainly not needed and won't ever be.

File: Source/WebKit/ModelProcess/ModelProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+#include <wtf/UniqueRef.h>
+#include <wtf/text/AtomString.h>
+
+namespace WebKit {
+
+// We wouldn't want the ModelProcess to repeatedly exit then relaunch when under memory pressure. In particular, we need to make sure the
+// WebProcess has a chance to schedule work after the ModelProcess gets launched. For this reason, we make sure that the ModelProcess never
+// idle-exits less than 5 seconds after getting launched. This amount of time should be sufficient for the WebProcess to schedule
+// work in the ModelProcess.
+constexpr Seconds minimumLifetimeBeforeIdleExit { 5_s };
```
Comment (achristensen07) [inline Source/WebKit/ModelProcess/ModelProcess.cpp:61]: This has a lot of duplicate code with the GPU process.  It would be much better to share code by maybe inheriting from a shared superclass rather than duplicate code.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj (hunk 5 above)
Comment (eddydas) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: This is great catch, thank you! Hid deep in xcodeproj.

File: Source/WebKit/ModelProcess/ModelProcess.cpp (hunk 9 above)
Comment (eddydas) [reply Source/WebKit/ModelProcess/ModelProcess.cpp:61]: This is indeed a feedback from a number of engineers too. We are going to explore that in the near future.

File: Source/WebKit/DerivedSources-input.xcfilelist (hunk 1 above)
Comment (eddydas) [reply Source/WebKit/DerivedSources-input.xcfilelist]: huh... it looks like the script won't remove it for me if I have removed it from DerivedSources.txt
will remove it manually

Comment (eddydas) [reply Source/WebKit/DerivedSources-input.xcfilelist]: removing WebKitBuild folder, resetting this file, and re-building => got this file to re-generate with just the required lines!

File: Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h [hunk 10]
```diff
@@ -27,6 +27,7 @@
 
 #if ENABLE(LEGACY_PDFKIT_PLUGIN)
 
+#include "PDFPluginBase.h"
```
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h]: This include shouldn't be necessary! Do you mind removing it on your next rebase?

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 11]
```diff
@@ -1055,6 +1056,12 @@ void WebProcessProxy::gpuProcessExited(ProcessTerminationReason reason)
 }
 #endif
 
+#if ENABLE(MODEL_PROCESS)
+void WebProcessProxy::createModelProcessConnection(IPC::Connection::Handle&& connectionIdentifier, WebKit::ModelProcessConnectionParameters&& parameters)
```
Comment (aprotyas) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: You can skip the argument names for now

**Review by emw-apple — COMMENTED** [review]
I'll be looking forward to the upcoming build architecture patch. We've talked already about the ExtensionKit migration that's underway; that's the only structural thing I can think of that makes it harder than just copy-pasting an existing XPC target.

File: Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h (hunk 10 above)
Comment (eddydas) [reply Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h]: Looks like it! Fixed

File: Source/WebKit/CMakeLists.txt [hunk 12]
```diff
@@ -361,6 +363,8 @@ set(WebKit_SERIALIZATION_IN_FILES
     GPUProcess/media/TrackPrivateRemoteConfiguration.serialization.in
     GPUProcess/media/VideoTrackPrivateRemoteConfiguration.serialization.in
 
+    ModelProcess/ModelProcessCreationParameters.serialization.in
```
Comment (emw-apple) [inline Source/WebKit/CMakeLists.txt:366]: It's not entirely clear to me what this will do or if it's needed, since I assume that you'll only be adding the model process to the Xcode-based builds. But if it's just here to make some imports happy, I think that's fine.

**Review by achristensen07 — APPROVED** [review]
